Update emails to point to new Grant Details page

jeffsmohan opened the pull request:
### Ticket #2644 
## Description
Updates the digest and assigned emails to link to our new Grant Details page when the flag is on, instead of directly to grants.gov. 

## Screenshots / Demo Video
<img width="1357" alt="Screenshot 2024-03-01 at 12 01 12 PM" src="https://github.com/usdigitalresponse/usdr-gost/assets/126250/e38f5136-cc4c-4076-aae6-d223b10d5fb1">
<img width="1357" alt="Screenshot 2024-03-01 at 12 04 17 PM" src="https://github.com/usdigitalresponse/usdr-gost/assets/126250/bb98ae46-7a82-4be7-8037-f6cb90f6a5e2">

## Testing
Testing this in dev is a lot of pretty fiddly effort, but here are the steps: 

1. Set up so emails actually get sent/captured in dev
   1. Create an ethereal email to work with: https://ethereal.email/ (nice, one-click creation of an ephemeral email account)
   2. Update the `NODEMAILER_*` environment variables in your `packages/server/.env` file from the values listed on your ethereal account
   3. Comment out `secure: true` in our nodemailer setup (Gmail uses a secure connection, but ethereal doesn't): https://github.com/usdigitalresponse/usdr-gost/blob/05f688e75df1e3fbf6698f88b49281dd117eb280/packages/server/src/lib/email/email-nodemailer.js#L36
   4. Rebuild your docker container so it picks up the new env variables (`docker compose down app` and `docker compose up -d`)
2. Trigger the emails manually via a node REPL
   - Digest email
     1. `docker compose exec app yarn node`
     2. ```node
           const emailService = require('./packages/server/src/lib/email');
           const knex = require('./packages/server/src/db/connection')
           const grants = await knex('grants').limit(3)
           emailService.sendGrantDigest({name: 'Test agency', matchedGrants: grants, recipients: ['test@example.com'], openDate: '2024-01-01'})
           ```
   - Assignment email
     1. Comment out the `assignee` variable in `sendGrantAssignedNotficationForAgency` and replace it with `const assignees = [{ email: 'test@example.com' }];`: https://github.com/usdigitalresponse/usdr-gost/blob/05f688e75df1e3fbf6698f88b49281dd117eb280/packages/server/src/lib/email.js#L221
     2. `docker compose exec app yarn node`
     3. ```node
           const emailService = require('./packages/server/src/lib/email');
           const knex = require('./packages/server/src/db/connection')
           const grant = await knex('grants').first();
           emailService.sendGrantAssignedEmail({ grantId: grant.grant_id, agencyIds: [1], userId: 1 });
           ```
3. Once the test email is sent, go to your "Messages" tab in your Ethereal mail account, and you should see the sent email listed there for viewing

### Automated and Unit Tests
- [x] Added Unit tests

### Manual tests for Reviewer
- [x] Added steps to test feature/functionality manually

## Checklist
- [x] Provided ticket and description
- [x] Provided screenshots/demo
- [x] Provided testing information
- [x] Provided adequate test coverage for all new code
- [x] Added PR reviewers

inaki reviewed (approved):
In order test the rebuild was not enough, but it works if you delete docker instance and setup a new instance with the env variable and the testing changes.

inaki commented:
Also looks good in mobile.

<img width="271" alt="Screenshot 2024-03-05 at 4 00 59 PM" src="https://github.com/usdigitalresponse/usdr-gost/assets/485547/d037089d-b940-41e3-80cc-c8e74ac6bad1">

jeffsmohan commented:
@TylerHendrickson Even though this has already been reviewed and approved, I'd appreciate your eyes on the code too, particularly the staging/prod flags.

TylerHendrickson commented on packages/server/src/lib/email.js:
```diff
@@ -218,10 +236,10 @@ async function sendGrantAssignedNotficationForAgency(assignee_agency, grantDetai
     // TODO: add plain text version of the email
     const emailPlain = emailHTML.replace(/<[^>]+>/g, '');
     const emailSubject = `Grant Assigned to ${assignee_agency.name}`;
-    const assginees = await db.getSubscribersForNotification(assignee_agency.id, notificationType.grantAssignment);
+    const assignees = await db.getSubscribersForNotification(assignee_agency.id, notificationType.grantAssignment);
 
     const inputs = [];
-    assginees.forEach((assignee) => inputs.push(
+    assignees.forEach((assignee) => inputs.push(
```
Good catch 🎩

TylerHendrickson commented on packages/server/src/static/email_templates/_grant_detail.html:
```diff
@@ -31,7 +31,7 @@
                                                 <p
                                                     style="Margin:0;-webkit-text-size-adjust:none;-ms-text-size-adjust:none;mso-line-height-rule:exactly;font-family:arial, 'helvetica neue', helvetica, sans-serif;line-height:24px;color:#000000;font-size:16px">
-                                                    {{{description}}}<span>... View on <a style="color:#0068D6;text-decoration:none;" href="{{link_url}}">Grants.gov</a></span>
+                                                    {{{description}}}<span>... View on <a style="color:#0068D6;text-decoration:none;" href="{{link_url}}">{{ link_description }}</a></span>
```
Noticed that the `href` values are being URL-encoded...
```suggestion
                                                    {{{description}}}<span>... View on <a style="color:#0068D6;text-decoration:none;" href="{{& link_url }}">{{ link_description }}</a></span>
```
Would you mind updating this file to include `&` on any other `href` values provided by mustache?

jeffsmohan replied:
Good catch! 

I'm trying to think of a good way to prevent opening ourselves up to an injection attack here. (I think email clients by and large have a lot of guardrails to prevent stray scripts anyway, but I'd still prefer to avoid the issue on our end.) I think for now I'll adopt the convention of appending `_safe` to a variable name to indicate it's known to be safe, and escape the value from the DB when building the template variables. But if you have another preference here, let me know!

I'm noting we have the same tougher-to-solve issue with rendering the grant description un-escaped here as well, in addition to our use of `v-html` in frontend code. I've added a note to the ticket tracking that issue: https://github.com/usdigitalresponse/usdr-gost/issues/2572#issuecomment-1986096350

TylerHendrickson replied:
This solution looks good to me!

TylerHendrickson reviewed (changes requested):
Requesting a quick tweak, otherwise looks good!

jeffsmohan commented:
@TylerHendrickson Thanks for the review! I've addressed the URL escaping issue, and I also fixed another bug I found with an incorrect utm_campaign value.
